signer/metrics.go

```diff
@@ -0,0 +1,262 @@
+	}
+}
+
+func (mt *metricsTimer) SetPreviousPrecommit(t time.Time) {
+	mt.mu.Lock()
+	defer mt.mu.Unlock()
+	mt.previousPrecommit = t
+}
+
+func (mt *metricsTimer) SetPreviousPrevote(t time.Time) {
+	mt.mu.Lock()
+	defer mt.mu.Unlock()
+	mt.previousPrevote = t
+}
+
+func (mt *metricsTimer) SetPreviousLocalSignStart(t time.Time) {
+	mt.mu.Lock()
+	defer mt.mu.Unlock()
+	mt.previousLocalSignStart = t
+}
+
+func (mt *metricsTimer) SetPreviousLocalSignFinish(t time.Time) {
+	mt.mu.Lock()
+	defer mt.mu.Unlock()
+	mt.previousLocalSignFinish = t
+}
+
+func (mt *metricsTimer) SetPreviousLocalEphemeralShare(t time.Time) {
+	mt.mu.Lock()
+	defer mt.mu.Unlock()
+	mt.previousLocalEphemeralShare = t
+}
+
+func (mt *metricsTimer) UpdatePrometheusMetrics(t time.Time) {
+	mt.mu.Lock()
+	defer mt.mu.Unlock()
+
+	// Update Prometheus Gauges
+	secondsSinceLastPrecommit.Set(time.Since(mt.previousPrecommit).Seconds())
+	secondsSinceLastPrevote.Set(time.Since(mt.previousPrevote).Seconds())
+	secondsSinceLastLocalSignStart.Set(time.Since(mt.previousLocalSignStart).Seconds())
+	secondsSinceLastLocalSignFinish.Set(time.Since(mt.previousLocalSignFinish).Seconds())
+	secondsSinceLastLocalEphemeralShareTime.Set(time.Since(mt.previousLocalEphemeralShare).Seconds())
+}
+
+var (
+	// Variables to calculate Prometheus Metrics
+	previousPrecommitHeight = int64(0)
+	previousPrevoteHeight   = int64(0)
+	metricsTimeKeeper       = newMetricsTimer()
+
+	// Prometheus Metrics
+	totalPubKeyRequests = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "signer_total_pubkey_requests",
+		Help: "Total times public key requested (High count may indicate validator restarts)",
+	})
+	lastPrecommitHeight = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "signer_last_precommit_height",
+		Help: "Last Height Precommit Signed",
+	})
+	lastPrevoteHeight = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "signer_last_prevote_height",
+		Help: "Last Height Prevote Signed",
+	})
+
+	lastProposalHeight = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "signer_last_proposal_height",
+		Help: "Last Height Proposal Signed",
+	})
+	lastPrecommitRound = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "signer_last_precommit_round",
+		Help: "Last Round Precommit Signed",
+	})
+	lastPrevoteRound = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "signer_last_prevote_round",
+		Help: "Last Round Prevote Signed",
+	})
+	lastProposalRound = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "signer_last_proposal_round",
+		Help: "Last Round Proposal Signed",
+	})
+
+	totalPrecommitsSigned = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "signer_total_precommits_signed",
+		Help: "Total Precommit Signed",
+	})
+	totalPrevotesSigned = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "signer_total_prevotes_signed",
+		Help: "Total Prevote Signed",
+	})
+	totalProposalsSigned = promauto.NewGauge(prometheus.GaugeOpts{
+		Name: "signer_total_proposals_signed",
+		Help: "Total Proposal Signed",
+	})
```

Comment:
Should these gauges be counters instead?

Reply:
Good chance.  Let me check/adjust